CRI: upgrade protobuf to v3

feiskyer opened the pull request:
For #38854, this PR upgrades CRI protobuf version to v3, and also updated related packages for confirming to new api.  

**Release note**:

```
CRI: upgrade protobuf version to v3.
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/39158)
<!-- Reviewable:end -->

feiskyer commented:
@k8s-bot cri e2e test this

feiskyer commented:
cc @yujuhong @yifan-gu @euank

feiskyer commented:
@k8s-bot bazel test this

feiskyer commented:
All tests passed. cc/ @kubernetes/sig-node-misc

yifan-gu reviewed (commented):
Commit a4201841e7c021fa2771698939ded0f5e7b5a76e lgtm overall, with minor nits
Commit a9feac7897c0d38d001f2ff896d8286a13d4e158 lgtm overall
Some comments on Commit c52e402cb81443b9a2e739bd43d28218c3ac0c76 
Commit 19e4d20e9d2f3482703bea80065bff109ff460ef lgtm
Commit e441e7d3641f3a1c0add1019bac336eb2e0d689d lgtm
Commit bee119111ee55316b18de5d38f39eb1a9e4fdba9 lgtm with some nits
Commit e819c6f00fe293e111805dbf4ac3c525c6da4fcf lgtm.

Finished 1st round of review.

yifan-gu commented:
Grep'd the integer and boolean in the old proto from current HEAD, to make sure default value will make sense in new proto.

https://docs.google.com/document/d/1VOpUbimwKvtK1gqxiXZEMDgN1hJeuYmewXI3dCHmLYI/edit

yifan-gu commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -292,77 +292,78 @@ message PodSandboxConfig {
 enum PodSandboxState {
-    SANDBOX_READY = 0;
-    SANDBOX_NOTREADY = 1;
+    SANDBOX_NOTSPECIFIED = 0;
```
Actually I don't feel this is really necessary, the runtime should specify the state anyway before returning the status, and kubelet should be able to make assumption about that, right? @yujuhong @feiskyer

yujuhong replied:
Yes, I think this is too conservative -- `PodSandboxState` is not really optional.

feiskyer replied:
`SANDBOX_NOTSPECIFIED` is not required. will remove it.

yifan-gu commented on pkg/kubelet/api/testing/fake_runtime_service.go:
```diff
@@ -321,13 +317,13 @@ func (r *FakeRuntimeService) ListContainers(filter *runtimeapi.ContainerFilter)
 				continue
 			}
-			if filter.State != nil && filter.GetState() != s.GetState() {
+			if filter.State != runtimeapi.ContainerState_CONTAINER_NOTSPECIFIED && filter.State != s.State {
```
Per https://github.com/kubernetes/kubernetes/pull/39158#discussion_r96540658
If we decide remove the default value, we will simplify the code here as well.

yifan-gu commented:
commit https://github.com/kubernetes/kubernetes/pull/39158#discussion_r96540658 lgtm overall

yifan-gu commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -590,105 +591,106 @@ message ContainerConfig {
-    CONTAINER_RUNNING = 1;
-    CONTAINER_EXITED  = 2;
-    CONTAINER_UNKNOWN = 3;
+    CONTAINER_NOTSPECIFIED = 0;
```
ditto

yifan-gu commented on pkg/kubelet/kuberuntime/helpers.go:
```diff
@@ -120,34 +120,34 @@ func (m *kubeGenericRuntimeManager) toKubeContainer(c *runtimeapi.Container) (*k
+	imageStatus, err := m.imageService.ImageStatus(&runtimeapi.ImageSpec{Image: image})
 	if err != nil {
-		return nil, nil, err
+		return 0, "", err
```
nit: this shouldn't matter, but can you return a negative value e.g. -1 on failure, just for reading.

yifan-gu commented on pkg/kubelet/kuberuntime/helpers.go:
```diff
@@ -120,34 +120,34 @@ func (m *kubeGenericRuntimeManager) toKubeContainer(c *runtimeapi.Container) (*k
 	}
 
-	if imageStatus != nil && imageStatus.Uid != nil {
+	if imageStatus != nil && imageStatus.Uid != 0 {
```
This part of logic can be rewritten to:
```go
if imageStatus != nil {
    return imageStatus.Uid, imageStatus.Username, nil
}
return 0, "", nil
```

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -139,35 +139,35 @@ func (m *kubeGenericRuntimeManager) generateContainerConfig(container *v1.Contai
 		}
 	} else {
-		glog.Warningf("Non-root verification doesn't support non-numeric user (%s)", *username)
+		glog.Warningf("Non-root verification doesn't support non-numeric user (%s)", username)
```
the condition has changed, should be `else if username != ""`

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:
```diff
@@ -158,17 +157,16 @@ func (m *kubeGenericRuntimeManager) generatePodSandboxLinuxConfig(pod *v1.Pod, c
 	if kubecontainer.HasPrivilegedContainer(pod) {
-		privileged := true
-		lc.SecurityContext.Privileged = &privileged
+		lc.SecurityContext.Privileged = true
```
nit:
```go
lc.SecurityContext.Privileged = kubecontainer.HasPrivilegedContainer(pod)
```

yifan-gu commented on pkg/kubelet/dockershim/docker_container.go:
```diff
@@ -42,14 +42,14 @@ func (ds *dockerService) ListContainers(filter *runtimeapi.ContainerFilter) ([]*
 		}
-		if filter.State != nil {
-			f.Add("status", toDockerContainerStatus(filter.GetState()))
+		if filter.State != runtimeapi.ContainerState_CONTAINER_NOTSPECIFIED {
```
ditto, this default value is not used.

yifan-gu commented on pkg/kubelet/dockershim/docker_sandbox.go:
```diff
@@ -242,11 +242,11 @@ func (ds *dockerService) ListPodSandbox(filter *runtimeapi.PodSandboxFilter) ([]
 		}
-		if filter.State != nil {
-			if filter.GetState() == runtimeapi.PodSandboxState_SANDBOX_READY {
+		if filter.State != runtimeapi.PodSandboxState_SANDBOX_NOTSPECIFIED {
```
ditto about the default value.

feiskyer commented:
> The default values may cause problems for some integer fields. For example, ExitCode will be default to 0. We could use some message type:
https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.Int32Value

@yujuhong Reviewed codes and I think there is no problem of having default value `0` for integer fields. For example, [ContainerStatus.ExitCode](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/container/runtime.go#L321) has already been assumed default 0 if not set explicitly because it is defined as an `int`. Do you think there are other potential problems?



Addressed comments and moved state of container/sandbox in filters into wrappers.

@yifan-gu @yujuhong PTAL.

feiskyer commented:
@k8s-bot gce etcd3 e2e test this

feiskyer commented:
@k8s-bot kops aws e2e test this

yifan-gu commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -590,105 +596,111 @@ message ContainerConfig {
-    CONTAINER_RUNNING = 1;
-    CONTAINER_EXITED  = 2;
-    CONTAINER_UNKNOWN = 3;
+    CONTAINER_CREATED      = 0;
```
why the consts here have a different indent (extra empty spaces) than others?

feiskyer replied:
This is because of removing an extra state in the old commits. Will remove those extra spaces.

yifan-gu commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -292,77 +292,77 @@ message PodSandboxConfig {
 enum PodSandboxState {
-    SANDBOX_READY = 0;
-    SANDBOX_NOTREADY = 1;
+    SANDBOX_READY        = 0;
```
nit: why they have extra spaces?

yifan-gu commented:
small nits, LGTM from my side.

yifan-gu commented:
It's not from the PR, but can we also rename the `status` to `ready` ? `status` doesn't sound like a boolean
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/api/v1alpha1/runtime/api.proto#L925

cc @Random-Liu 

Also their is a typo in the comments: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/api/v1alpha1/runtime/api.proto#L910
(s/condtition/condition)

yujuhong commented:
Reviewed 1 of 20 files at r2.
Review status: 1 of 43 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/kubelet/api/v1alpha1/runtime/api.proto, line 445 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39158#-Kan-IIzPYvH8Dj0U_3Q:-Kan-IIzPYvH8Dj0U_3R:bolk6ar) ([raw file](https://github.com/kubernetes/kubernetes/blob/d92d582d17a1f43401ba87e002e87b880ccc33c9/pkg/kubelet/api/v1alpha1/runtime/api.proto#L445)):*
> ```protobuf
> // TODO: Consider using Resources from opencontainers/runtime-spec/specs-go
> // directly.
> message LinuxContainerResources {
> ```

Could you check if it's ok to use 0 as the default for these fields?

---

*[pkg/kubelet/api/v1alpha1/runtime/api.proto, line 502 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39158#-Kan8cHWV5jbnoRYiCU3:-Kan8cHWV5jbnoRYiCU4:b-85racv) ([raw file](https://github.com/kubernetes/kubernetes/blob/d92d582d17a1f43401ba87e002e87b880ccc33c9/pkg/kubelet/api/v1alpha1/runtime/api.proto#L502)):*
> ```protobuf
>     SELinuxOption selinux_options = 4;
>     // UID to run the container process as. Only one of run_as_user and
>     // run_as_username can be specified at a time.
> ```

How do we know if `run_as_user` is set? 

---

*[pkg/kubelet/api/v1alpha1/runtime/api.proto, line 641 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39158#-Kan-xLqOST1kGR6QVoq:-Kan-xLqOST1kGR6QVor:b3cypnt) ([raw file](https://github.com/kubernetes/kubernetes/blob/d92d582d17a1f43401ba87e002e87b880ccc33c9/pkg/kubelet/api/v1alpha1/runtime/api.proto#L641)):*
> ```protobuf
>     string container_id = 1;
>     // Timeout, in seconds, to stop the container.
>     int64 timeout = 2;
> ```

Do we allow not specifying timeout? 

---

*[pkg/kubelet/api/v1alpha1/runtime/api.proto, line 771 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39158#-Kan90DboLAU549w79h7:-Kan90DboLAU549w79h8:b-5blrp7) ([raw file](https://github.com/kubernetes/kubernetes/blob/d92d582d17a1f43401ba87e002e87b880ccc33c9/pkg/kubelet/api/v1alpha1/runtime/api.proto#L771)):*
> ```protobuf
>     repeated string cmd = 2;
>     // Timeout in seconds to stop the command. Default: run forever.
>     int64 timeout = 3;
> ```

Can we rely on timeout=0 to indicate running forever?

---

*[pkg/kubelet/api/v1alpha1/runtime/api.proto, line 849 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39158#-Kan9M_bm17BAOE_IL_A:-Kan9M_bm17BAOE_IL_B:b-axfpmg) ([raw file](https://github.com/kubernetes/kubernetes/blob/d92d582d17a1f43401ba87e002e87b880ccc33c9/pkg/kubelet/api/v1alpha1/runtime/api.proto#L849)):*
> ```protobuf
>     // specified when creating the container. UID and the following user name
>     // are mutually exclusive.
>     int64 uid = 5;
> ```

How do we know if uid is not set or actually 0?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39158)*
<!-- Sent from Reviewable.io -->

yujuhong commented:
I have questions about some of the integer-type fields and their default value being 0. Will review the rest once these are addressed.

yifan-gu commented:
LGTM overall, the default 0 values are passed to runtime by GetXXX() today already, e.g. GetTimeout(), so there is no functional changes happening here.

But we should better document all those integer default values in the proto. [checklist here](https://docs.google.com/document/d/1VOpUbimwKvtK1gqxiXZEMDgN1hJeuYmewXI3dCHmLYI/edit)

yujuhong commented:
> The default 0 values are passed to runtime by GetXXX() today already, e.g. GetTimeout(), so there is no functional changes happening here.

Better document is needed for sure, especially now that there is no way to check the existence of these field even if we want to. The use of timeout is not very consistent -- in one case (`StopContainer) it's used to indicate immediate termination, while in the other (`ExecSync` request) it means run forever. What 0 means for the resource quota/limit fields is just unclear.
 
As for `run_as_user` and `uid`, we do check them in the code, and I don't think we can not wrap them in proto3.

Random-Liu commented:
> t's not from the PR, but can we also rename the status to ready ? status doesn't sound like a boolean

I name it as `status` just to match the K8s api https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L1371-L1381.

"ready" or not should be in condition type (`RuntimeReady`, `NetworkNotReady`), status should only indicate whether the condition is true or not.

feiskyer commented:
@yujuhong @yifan-gu Addressing comments. PTAL.

yujuhong commented:
Reviewed 17 of 41 files at r1, 15 of 20 files at r2, 11 of 11 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[hack/verify-flags/exceptions.txt, line 70 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39158#-KarmDGaBg1Q5goOxIQR:-KarmDGbabSwzZsK7Ew3:b-j1ypsb) ([raw file](https://github.com/kubernetes/kubernetes/blob/d357a7216106095ca885e6a437b05ed3071289b3/hack/verify-flags/exceptions.txt#L70)):*
> ```
> cluster/saltbase/salt/opencontrail-networking-master/init.sls:      - 'SERVICE_CLUSTER_IP_RANGE': '{{ pillar.get('service_cluster_ip_range') }}'
> cluster/saltbase/salt/opencontrail-networking-minion/init.sls:      - 'SERVICE_CLUSTER_IP_RANGE': '{{ pillar.get('service_cluster_ip_range') }}'
> cluster/saltbase/salt/supervisor/kubelet-checker.sh:	{% set kubelet_port = pillar['kubelet_port'] -%}
> ```

Why would the PR touch this?...

---

*[pkg/kubelet/api/v1alpha1/runtime/api.proto, line 647 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39158#-Kark82zgYpZkp8bUHQw:-Kark830MWtkL23K3UMR:bkd4tr) ([raw file](https://github.com/kubernetes/kubernetes/blob/34304fdc2bc00e1ddaa417afe92f36b308953ffc/pkg/kubelet/api/v1alpha1/runtime/api.proto#L647)):*
> ```protobuf
>     string container_id = 1;
>     // Timeout in seconds to stop the container. Must be > 0.
>     int64 timeout = 2;
> ```

I think we should allow timeout=0. How about:
`Timeout in seconds to wait for the container to stop before forcibly terminating it. Default: 0 (forcibly terminate the container immediately)`

---

*[pkg/kubelet/api/v1alpha1/runtime/api.proto, line 823 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39158#-KarkSq8j1gvGVmvrRph:-KarkSq8j1gvGVmvrRpi:bd40je2) ([raw file](https://github.com/kubernetes/kubernetes/blob/34304fdc2bc00e1ddaa417afe92f36b308953ffc/pkg/kubelet/api/v1alpha1/runtime/api.proto#L823)):*
> ```protobuf
>     // ID of the container to which to forward the port.
>     string pod_sandbox_id = 1;
>     // Port to forward. Each port must be > 0 if provided.
> ```

nit: I think `repeated` fields should be ok because they'll get converted to a list (which could be nil).

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39158)*
<!-- Sent from Reviewable.io -->

yujuhong commented:
Looks good overall with some nits. 
Probably also want to squash some of the commits.

yifan-gu commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -699,7 +705,7 @@ message Container {
     // State of the container.
     ContainerState state = 6;
-    // Creation time of the container in nanoseconds.
+    // Creation time of the container in nanoseconds. Must be > 0.
```
while the container is not created, this can be 0 right?

yifan-gu commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -728,13 +734,13 @@ message ContainerStatus {
     // Status of the container.
     ContainerState state = 3;
-    // Creation time of the container in nanoseconds.
+    // Creation time of the container in nanoseconds. Must be > 0.
```
ditto

yifan-gu commented:
@Random-Liu Thanks for the clarification.

@feiskyer LGTM except the comments made by @yujuhong

feiskyer committed:
CRI: update protobuf to v3

feiskyer committed:
kubelet/api: update cri to protobuf v3

feiskyer committed:
kubelet/server: update cri to protobuf v3

feiskyer committed:
kubelet/remote: update cri to protobuf v3

feiskyer committed:
kubelet/rktshim: update cri to protobuf v3

feiskyer committed:
kubelet/dockershim: update cri to protobuf v3

feiskyer committed:
kubelet/kuberuntime: update cri to protobuf v3

feiskyer committed:
Update flags underscore

feiskyer commented:
Rebased and squashed commits.

> Why would the PR touch this?

`hack/verify-flags/exceptions.txt` is generated automatically by `hack/verify-flags-underscore.py -e`.

yujuhong commented:
/lgtm

Please help watch the [cri builds](https://k8s-testgrid.appspot.com/google-cri) to see if anything breaks after the PR is merged.

feiskyer commented:
> Please help watch the cri builds to see if anything breaks after the PR is merged.

Of course. Is CRI enabled for all tests in page [google-cri](https://k8s-testgrid.appspot.com/google-cri) ?

Random-Liu commented:
> Of course. Is CRI enabled for all tests in page google-cri ?

Yeah.

yujuhong commented:
FYI, I glanced over all the cri builds and didn't see any obvious red builds.
